### ISSUE TITLE
Support Windows emacs with WSL server

### ISSUE THREAD

**Issue by garyo** [emacs-lsp/lsp-mode#761]
I'm on Windows 10, running native emacs 26.1. I'm doing some Javascript/Typescript (vue.js) development, using WSL for that world, still with the native Windows Emacs. Works great for the most part, with a few path hassles now & then (/mnt/c/foo vs. c:/foo). I definitely know it's a fringe use case, so I'll understand if you can't help.

I just ran across the LSP idea -- brilliant! I want to try that to get nice intellisense. I'll need to run the LSP server on WSL though. I spent some time on it.
The first thing is to start the server. I created a .cmd file and use that as the server:
```bat
% cat /mnt/c/bin/typescript-language-server.cmd
@echo off
c:/windows/system32/bash.exe -c "/path-to-npm-global/typescript-language-server --stdio"
```

But then the big thing that needs tweaking is the file and URI paths. On Windows, the file is `c:/foo/bar.ts` but in WSL that needs to be translated to `/mnt/c/foo/bar.ts` and similarly for the URIs. 
I created a simple patch to do this (this is just proof of concept to see if I can get it working), and it does enable the server to start.

```patch
*** /mnt/c/tmp/lsp-mode.el.orig 2019-04-12 09:38:03.376623200 -0400
--- lsp-mode.el 2019-04-12 09:53:53.206487300 -0400
***************
*** 696,701 ****
--- 696,707 ----
        "Return the local name component of FILE."
        (or (file-remote-p file 'localname) file))))

+ (defun lsp-server-path (path)
+   (replace-regexp-in-string "^c:/" "/mnt/c/" path))
+
+ (defun lsp-server-path-for-uri (path)
+   (replace-regexp-in-string "^c:/" "mnt/c/" path))
+
  (defun lsp--merge-results (results method)
    "Merge RESULTS by filtering the empty hash-tables and merging the lists.
  METHOD is the executed method so the results could be merged
***************
*** 976,982 ****
  (defun lsp--path-to-uri (path)
    "Convert PATH to a uri."
    (concat lsp--uri-file-prefix
!           (url-hexify-string (expand-file-name (or (file-remote-p path 'localname t) path))
                               url-path-allowed-chars)))

  (define-inline lsp--string-match-any (regex-list str)
--- 982,989 ----
  (defun lsp--path-to-uri (path)
    "Convert PATH to a uri."
    (concat lsp--uri-file-prefix
!           (url-hexify-string (lsp-server-path-for-uri
!                               (expand-file-name (or (file-remote-p path 'localname t) path)))
                               url-path-allowed-chars)))

  (define-inline lsp--string-match-any (regex-list str)
***************
*** 4401,4407 ****
        (lsp-request-async "initialize"
                           (append
                            (list :processId (emacs-pid)
!                                 :rootPath (lsp-file-local-name (expand-file-name root))
                                  :rootUri (lsp--path-to-uri root)
                                  :capabilities (lsp--client-capabilities)
                                  :initializationOptions initialization-options)
--- 4408,4414 ----
        (lsp-request-async "initialize"
                           (append
                            (list :processId (emacs-pid)
!                                 :rootPath (lsp-server-path (lsp-file-local-name (expand-file-name root)))
                                  :rootUri (lsp--path-to-uri root)
                                  :capabilities (lsp--client-capabilities)
                                  :initializationOptions initialization-options)
```

But after the server starts and exchanges a couple of messages, it seems to exit silently with status 1, and I'm not sure why. The stdout and stderr buffers are empty, and the log just shows the client sending a few messages and getting no reply (consistent with the server exiting).
```
LSP :: Connected to [ts-ls:17708 status:starting].
LSP :: ts-ls has exited (exited abnormally with code 1)
Server ts-ls:17708 exited with status exit.
```

Here's the log (thanks for enabling all this tracing functionality, btw -- super helpful!)
You can see the server sends the first valid response, but then nothing after that.

```
Command "javascript-typescript-stdio" is not present on the path.
Command "c:/bin/typescript-language-server.cmd" is present on the path.
Found the following clients for .../project.module.ts: (server-id ts-ls, priority -2)
The following clients were selected based on priority: (server-id ts-ls, priority -2)
>>> ts-ls:17708 status:starting(async)
Content-Length: 3202

{
  "jsonrpc": "2.0",
  "method": "initialize",
  "params": {
    "processId": 21176,
    "rootPath": "/mnt/c/dss/Product/Horizon/WebProjects/horizon-project/horizon/",
    "rootUri": "file:///mnt/c/dss/Product/Horizon/WebProjects/horizon-project/horizon/",
    "capabilities": {
      "workspace": {
        "workspaceEdit": {
          "documentChanges": true,
          "resourceOperations": [
            "create",
            "rename",
            "delete"
          ]
        },
        "applyEdit": true,
        "symbol": {
          "symbolKind": {
            "valueSet": [
              1,
              ...
            ]
          }
        },
        "executeCommand": {
          "dynamicRegistration": false
        },
        "didChangeWatchedFiles": {
          "dynamicRegistration": true
        },
        "workspaceFolders": true,
        "configuration": true
      },
      "textDocument": {
        "declaration": {
          "linkSupport": true
        },
        "definition": {
          "linkSupport": true
        },
        "implementation": {
          "linkSupport": true
        },
        "typeDefinition": {
          "linkSupport": true
        },
        "synchronization": {
          "willSave": true,
          "didSave": true,
          "willSaveWaitUntil": true
        },
        "documentSymbol": {
          "symbolKind": {
            "valueSet": [
              1,
              ...
            ]
          },
          "hierarchicalDocumentSymbolSupport": true
        },
        "formatting": {
          "dynamicRegistration": true
        },
        "codeAction": {
          "dynamicRegistration": true
        },
        "completion": {
          "completionItem": {
            "snippetSupport": true
          }
        },
        "signatureHelp": {
          "signatureInformation": {
            "parameterInformation": {
              "labelOffsetSupport": true
            }
          }
        },
        "documentLink": {
          "dynamicRegistration": true
        },
        "hover": {
          "contentFormat": [
            "plaintext",
            "markdown"
          ]
        },
        "foldingRange": {
          "dynamicRegistration": true,
          "rangeLimit": null,
          "lineFoldingOnly": null
        }
      }
    },
    "initializationOptions": null
  },
  "id": 1
}

<<<< ts-ls:17708 status:starting
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "capabilities": {
      "textDocumentSync": 2,
      "completionProvider": {
        "triggerCharacters": [
          ".",
          "\"",
          "'",
          "/",
          "@",
          "<"
        ],
        "resolveProvider": true
      },
      "codeActionProvider": true,
      "definitionProvider": true,
      "documentFormattingProvider": true,
      "documentHighlightProvider": true,
      "documentSymbolProvider": true,
      "executeCommandProvider": {
        "commands": [
          "_typescript.applyWorkspaceEdit",
          "_typescript.applyCodeAction",
          "_typescript.applyRefactoring",
          "_typescript.organizeImports",
          "_typescript.applyRenameFile"
        ]
      },
      "hoverProvider": true,
      "renameProvider": true,
      "referencesProvider": true,
      "signatureHelpProvider": {
        "triggerCharacters": [
          "(",
          ",",
          "<"
        ]
      },
      "workspaceSymbolProvider": true,
      "implementationProvider": true,
      "typeDefinitionProvider": true,
      "foldingRangeProvider": true,
      "callsProvider": true
    }
  }
}
>>> ts-ls:17708(async)
Content-Length: 69

{
  "jsonrpc": "2.0",
  "method": "initialized",
  "params": {
  }
}

>>> ts-ls:17708(async)
Content-Length: 1965

{
  "jsonrpc": "2.0",
  "method": "textDocument/didOpen",
  "params": {
    "textDocument": {
      "uri": "file:///mnt/c/dss/.../project.module.ts",
      "languageId": "typescript",
      "version": 0,
      "text": "// Vuex store..."
    }
  }
}

>>> ts-ls:17708(async)
Content-Length: 291

{
  "jsonrpc": "2.0",
  "method": "textDocument/hover",
  "params": {
    "textDocument": {
      "uri": "file:///mnt/c/dss/Product/Horizon/WebProjects/horizon-project/horizon/src/store/project.module.ts"
    },
    "position": {
      "line": 0,
      "character": 0
    }
  },
  "id": 2
}

>>> ts-ls:17708(async)
Content-Length: 437

{
  "jsonrpc": "2.0",
  "method": "textDocument/codeAction",
  "params": {
    "textDocument": {
      "uri": "file:///mnt/c/dss/Product/Horizon/WebProjects/horizon-project/horizon/src/store/project.module.ts"
    },
    "range": {
      "start": {
        "line": 0,
        "character": 0
      },
      "end": {
        "line": 0,
        "character": 30
      }
    },
    "context": {
      "diagnostics": []
    }
  },
  "id": 3
}
```

**Comment by yyoncho** [comment]
It is really hard to tell what is happening since it is in the language server. I think that I spotted one problem, linux paths are prefixed via two `//` (see https://github.com/emacs-lsp/lsp-mode/blob/master/lsp-mode.el#L184 )

As a side note, if it is possible to access wsl via ssh you may use tramp to edit the files which is supported. (lsp-mode supports running servers over tramp).


(cc @madnificent who wants to implement something like that in lsp-mode).

**Comment by madnificent** [comment]
👋 glad I'm not the only one with the strange idea of path name translation ;-)

@yyoncho Would a hook for back and forth translation of paths make sense in this context?

**Comment by yyoncho** [comment]
It should not be a hook but something that we register in lsp-client. Note that you may have several type of project topologies and you should be able to configure that per client/workspace.

**Comment by yyoncho** [comment]
@garyo any update on that, did you had a chance to try what was suggested in https://github.com/emacs-lsp/lsp-mode/issues/761#issuecomment-482634759 ?

**Comment by yyoncho** [comment]
fixed with 697d319
